File: ios/KoberletCore/Sources/KoberletCore/Carteras.swift
// Koberlet iOS - Copyright 2026 DNNS.es (Oberluss)
// SPDX-License-Identifier: Apache-2.0

import Foundation

/// EL CONTENIDO DE LA BOVEDA - varias carteras en el mismo aparato.
///
/// Formato v3 - UNA CARTERA, UNA RED (el mismo que Android y escritorio):
///   { v:3, carteras:[ { id, etiqueta, semilla|privada, red:"kda"|"evm", cuentas:[una] } ],
///     activa:"id" }
///
/// Los formatos v1 y v2 se migran al abrir, igual que en Carteras.kt: una copia
/// hecha en el Android viejo tiene que abrirse aqui sin perder nada.
///
/// A diferencia de Kotlin, los diccionarios de Swift son valores: cada funcion
/// DEVUELVE el contenido modificado y quien llama tiene que quedarse con lo que
/// devuelve, no con lo que le paso.
public enum Carteras {

    public static let VERSION = 3

    static func version(_ d: JSON) -> Int { (d["v"] as? Int) ?? 1 }
    static func lista(_ d: JSON) -> [JSON] { (d["carteras"] as? [JSON]) ?? [] }
    static func texto(_ d: JSON, _ clave: String) -> String { (d[clave] as? String) ?? "" }

    /// Devuelve el contenido en el formato de hoy, migrando si hace falta.
    public static func normalizar(_ datos: JSON) -> JSON { partirPorRed(aVarias(datos)) }

    /// v1 -> v2: la semilla suelta pasa a ser la primera cartera.
    private static func aVarias(_ datos: JSON) -> JSON {
        if version(datos) >= 2 && datos["carteras"] != nil { return datos }
        let cartera: JSON = [
            "id": "c1", "etiqueta": "Mi cartera",
            "semilla": texto(datos, "semilla"),
            "cuentas": (datos["cuentas"] as? [JSON]) ?? [],
        ]
        return ["v": 2, "carteras": [cartera], "activa": "c1"]
    }

    /// v2 -> v3: la cartera con dos cuentas se PARTE EN DOS, «X KDA» y «X EVM»,
    /// compartiendo la misma semilla. La de Kadena se queda con el id original.
    private static func partirPorRed(_ datos: JSON) -> JSON {
        if version(datos) >= VERSION { return datos }
        let viejas = lista(datos)
        var nuevas = [JSON]()
        var usados = Set<String>()
        for c in viejas { usados.insert(texto(c, "id")) }

        for c in viejas {
            let cuentas = (c["cuentas"] as? [JSON]) ?? []
            var porTipo = [String: JSON]()
            for cu in cuentas { porTipo[texto(cu, "tipo")] = cu }
            if porTipo.count <= 1 {
                var copia = c
                copia["red"] = porTipo.keys.first ?? "kda"
                nuevas.append(copia)
                continue
            }
            let nombre = texto(c, "etiqueta")
            var primera = true
            for tipo in ["kda", "evm"] {
                guard var cu = porTipo[tipo] else { continue }
                let id = primera ? texto(c, "id") : libre(usados)
                primera = false
                usados.insert(id)
                cu["id"] = "\(id)-\(tipo)"
                nuevas.append([
                    "id": id, "etiqueta": nombre + " " + tipo.uppercased(),
                    "semilla": texto(c, "semilla"), "red": tipo, "cuentas": [cu],
                ])
            }
        }
        let activa = texto(datos, "activa")
        let activaFinal = nuevas.isEmpty ? "" : (activa.isEmpty ? texto(nuevas[0], "id") : activa)
        return ["v": VERSION, "carteras": nuevas, "activa": activaFinal]
    }

    private static func libre(_ usados: Set<String>) -> String {
        var n = 1
        while usados.contains("c\(n)") { n += 1 }
        return "c\(n)"
    }

    public static func lista(de datos: JSON) -> [JSON] { lista(normalizar(datos)) }

    public static func buscar(_ datos: JSON, _ id: String) -> JSON? {
        lista(de: datos).first { texto($0, "id") == id }
    }

    /// La cartera activa, o la primera si la marcada ya no existe.
    public static func activa(_ datos: JSON) -> JSON? {
        let d = normalizar(datos)
        return buscar(d, texto(d, "activa")) ?? lista(d).first
    }

    /// Identificador libre: c1, c2, c3... Nunca se reutiliza uno ya usado.
    public static func idNuevo(_ datos: JSON) -> String {
        libre(Set(lista(de: datos).map { texto($0, "id") }))
    }

    public static func anadir(_ datos: JSON, _ cartera: JSON) -> JSON {
        var d = normalizar(datos)
        var l = lista(d)
        l.append(cartera)
        d["carteras"] = l
        d["activa"] = texto(cartera, "id")       // la recien creada pasa a ser la activa
        return d
    }

    /// Quita una cartera. Se niega a quitar la ultima: para empezar de cero esta
    /// `borrarTodo`, que avisa de lo que hace.
    public static func quitar(_ datos: JSON, _ id: String) throws -> JSON {
        var d = normalizar(datos)
        let l = lista(d)
        if l.count <= 1 { throw FalloBoveda.estado("Es la única cartera que hay. Para empezar de cero, usa borrar todo.") }
        let nueva = l.filter { texto($0, "id") != id }
        if nueva.count == l.count { throw FalloBoveda.argumento("Esa cartera no existe.") }
        d["carteras"] = nueva
        if texto(d, "activa") == id { d["activa"] = texto(nueva[0], "id") }
        return d
    }

    public static func renombrar(_ datos: JSON, _ id: String, _ etiqueta: String) throws -> JSON {
        var d = normalizar(datos)
        var l = lista(d)
        guard let i = l.firstIndex(where: { texto($0, "id") == id }) else {
            throw FalloBoveda.argumento("Esa cartera no existe.")
        }
        let limpia = String(etiqueta.trimmingCharacters(in: .whitespacesAndNewlines).prefix(40))
        if limpia.isEmpty { throw FalloBoveda.argumento("Ponle un nombre.") }
        l[i]["etiqueta"] = limpia
        d["carteras"] = l
        return d
    }

    /// ¿Ya hay una cartera con esta misma semilla PARA ESA RED?
    public static func yaExiste(_ datos: JSON, _ semilla: String, _ red: String) -> Bool {
        lista(de: datos).contains { texto($0, "semilla") == semilla && redDe($0) == red }
    }

    /// ¿Ya esta metida esta clave privada para esa red?
    public static func yaExisteClave(_ datos: JSON, _ privada: String, _ red: String) -> Bool {
        lista(de: datos).contains { texto($0, "privada") == privada && redDe($0) == red }
    }

    /// ¿Hay ya una cartera con ESA MISMA direccion?
    public static func yaExisteCuenta(_ datos: JSON, _ cuenta: String) -> Bool {
        for c in lista(de: datos) {
            for cu in (c["cuentas"] as? [JSON]) ?? [] where texto(cu, "cuenta") == cuenta { return true }
        }
        return false
    }

    /// ¿Esta cartera viene de unas palabras, o de una clave suelta?
    public static func esDeSemilla(_ cartera: JSON) -> Bool { !texto(cartera, "semilla").isEmpty }

    /// Deja la clave privada como se guarda: 64 caracteres hex en minusculas.
    /// Admite el `0x` delante y el formato de 128 caracteres (privada + publica)
    /// de algunas herramientas de Kadena, solo si la publica cuadra de verdad.
    public static func normalizarClave(_ texto: String, _ red: String) throws -> String {
        var h = texto.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        if h.hasPrefix("0x") { h = String(h.dropFirst(2)) }
        if !h.casa("^[0-9a-f]+$") {
            throw FalloBoveda.argumento("Una clave privada son 64 caracteres del 0 al 9 y de la a a la f.")
        }
        if h.count == 128 && red == "kda" {
            let privada = String(h.prefix(64))
            let publica = String(h.suffix(64))
            let calculada = try Derivacion.publicaKadena(try Hex.deHex(privada))
            if calculada != publica {
                throw FalloBoveda.argumento("Esos 128 caracteres no son una clave privada seguida de su pública.")
            }
            h = privada
        }
        if h.count != 64 {
            throw FalloBoveda.argumento("Una clave privada son 64 caracteres del 0 al 9 y de la a a la f.")
        }
        if h == String(repeating: "0", count: 64) { throw FalloBoveda.argumento("Esa clave no vale: son todo ceros.") }
        return h
    }

    /// Monta una cartera a partir de una clave privada suelta, sin semilla.
    public static func montarConClave(_ id: String, _ etiqueta: String, _ privada: String, _ red: String) throws -> JSON {
        if red != "kda" && red != "evm" { throw FalloBoveda.argumento("Esa red no existe.") }
        var bytes = try Hex.deHex(privada)
        defer { for i in bytes.indices { bytes[i] = 0 } }
        let cuenta: JSON
        if red == "kda" {
            cuenta = ["id": "\(id)-kda", "etiqueta": "Kadena", "tipo": "kda",
                      "cuenta": "k:" + (try Derivacion.publicaKadena(bytes))]
        } else {
            cuenta = ["id": "\(id)-evm", "etiqueta": "EVM", "tipo": "evm",
                      "cuenta": try Derivacion.direccionEvm(bytes)]
        }
        return ["id": id, "etiqueta": etiqueta, "privada": privada, "red": red, "cuentas": [cuenta]]
    }

    /// Los 32 bytes de la clave privada de esa cartera, venga de donde venga.
    /// Quien la pida se la lleva en un array que puede -y debe- borrar despues.
    public static func privadaDe(_ cartera: JSON) throws -> [UInt8] {
        let suelta = texto(cartera, "privada")
        if !suelta.isEmpty { return try Hex.deHex(suelta) }
        var bytes = Derivacion.semillaABytes(texto(cartera, "semilla"))
        defer { for i in bytes.indices { bytes[i] = 0 } }
        return redDe(cartera) == "kda"
            ? Derivacion.privadaKadena(bytes, 0)
            : try Derivacion.privadaEvm(bytes, 0)
    }

    /// La red de una cartera. Las de antes de la v3 no la traen: se deduce.
    public static func redDe(_ cartera: JSON) -> String {
        let r = texto(cartera, "red")
        if r == "kda" || r == "evm" { return r }
        guard let cuentas = cartera["cuentas"] as? [JSON], let primera = cuentas.first else { return "kda" }
        let tipo = texto(primera, "tipo")
        return tipo.isEmpty ? "kda" : tipo
    }

    /// Monta una cartera nueva a partir de una semilla, derivando SU cuenta.
    public static func montar(_ id: String, _ etiqueta: String, _ semilla: String, _ red: String) throws -> JSON {
        if red != "kda" && red != "evm" { throw FalloBoveda.argumento("Esa red no existe.") }
        var bytes = Derivacion.semillaABytes(semilla)
        defer { for i in bytes.indices { bytes[i] = 0 } }
        let cuenta: JSON
        if red == "kda" {
            cuenta = ["id": "\(id)-kda", "etiqueta": "Kadena", "tipo": "kda",
                      "cuenta": try Derivacion.cuentaKadena(bytes, 0)]
        } else {
            cuenta = ["id": "\(id)-evm", "etiqueta": "EVM", "tipo": "evm",
                      "cuenta": try Derivacion.direccionEvm(try Derivacion.privadaEvm(bytes, 0))]
        }
        return ["id": id, "etiqueta": etiqueta, "semilla": semilla, "red": red, "cuentas": [cuenta]]
    }
}

File: android/app/src/main/java/es/dnns/koberlet/Carteras.kt
// Koberlet Android - Copyright 2026 DNNS.es (Oberluss)
// SPDX-License-Identifier: Apache-2.0

package es.dnns.koberlet

import org.json.JSONArray
import org.json.JSONObject

/**
 * EL CONTENIDO DE LA BOVEDA - varias carteras en el mismo aparato.
 *
 * Formato v3 - UNA CARTERA, UNA RED:
 *   { v:3, carteras:[ { id, etiqueta, semilla, red:"kda"|"evm", cuentas:[una] } ],
 *     activa:"id" }
 *
 * Antes (v2) cada cartera derivaba de su semilla DOS cuentas, una de Kadena y
 * otra de Ethereum, y se enseñaban juntas. Es tecnicamente correcto -son dos
 * caminos BIP-44 de la misma semilla, `m/44'/626'/0'` y `m/44'/60'/0'/0/0`- pero
 * como modelo mental es malo: quien tiene tres carteras ve seis, y en el Puente
 * hay que elegir dos cosas que la app presentaba como una sola. El Koberlet de
 * escritorio ya lo cambio en julio de 2026 por el mismo motivo; esto alinea el
 * movil con aquello.
 *
 * Formatos anteriores:
 *   v1: { v:1, semilla:"...", cuentas:[...] }              (una sola semilla suelta)
 *   v2: { v:2, carteras:[{id,etiqueta,semilla,cuentas:[kda,evm]}], activa }
 *
 * Los dos **se migran al abrir**, sin perder nada. Esto importa mas de lo que
 * parece: quien ya tenga una version vieja instalada con su cartera dentro no
 * puede encontrarse con que la nueva "no ve" su dinero. Por eso la migracion
 * tiene sus propios tests.
 *
 * Esta clase no toca ficheros ni Android: es solo la forma de los datos, para
 * poder probarla entera en el ordenador.
 */
object Carteras {

    const val VERSION = 3

    /** Devuelve el contenido en el formato de hoy, migrando si hace falta. */
    fun normalizar(datos: JSONObject): JSONObject = partirPorRed(aVarias(datos))

    /** v1 -> v2: la semilla suelta pasa a ser la primera cartera. */
    private fun aVarias(datos: JSONObject): JSONObject {
        if (datos.optInt("v", 1) >= 2 && datos.has("carteras")) return datos
        val cartera = JSONObject()
            .put("id", "c1")
            .put("etiqueta", "Mi cartera")
            .put("semilla", datos.getString("semilla"))
            .put("cuentas", datos.optJSONArray("cuentas") ?: JSONArray())
        return JSONObject()
            .put("v", 2)
            .put("carteras", JSONArray().put(cartera))
            .put("activa", "c1")
    }

    /**
     * v2 -> v3: la cartera con dos cuentas se PARTE EN DOS, «X KDA» y «X EVM»,
     * compartiendo la misma semilla.
     *
     * No se pierde ni se mueve nada: las direcciones son exactamente las que ya
     * habia, porque salen de la misma semilla por el mismo camino. Lo unico que
     * cambia es como se agrupan.
     *
     * La de Kadena se queda con el id de la cartera original -asi la que estaba
     * marcada como activa, y la preferencia guardada en el aparato, siguen
     * valiendo- y la de Ethereum estrena id. El id de cada cuenta se rehace para
     * que siga siendo `<carteraId>-<tipo>`: de esa forma depende el exportar
     * claves privadas.
     */
    private fun partirPorRed(datos: JSONObject): JSONObject {
        if (datos.optInt("v", 1) >= VERSION) return datos
        val viejas = datos.getJSONArray("carteras")
        val nuevas = JSONArray()
        val usados = mutableSetOf<String>()
        for (i in 0 until viejas.length()) usados.add(viejas.getJSONObject(i).getString("id"))

        for (i in 0 until viejas.length()) {
            val c = viejas.getJSONObject(i)
            val cuentas = c.optJSONArray("cuentas") ?: JSONArray()
            val porTipo = mutableMapOf<String, JSONObject>()
            for (j in 0 until cuentas.length()) {
                val cu = cuentas.getJSONObject(j)
                porTipo[cu.optString("tipo")] = cu
            }
            // Una cartera con una sola cuenta ya es de una red: solo hay que decirlo.
            if (porTipo.size <= 1) {
                val tipo = porTipo.keys.firstOrNull() ?: "kda"
                nuevas.put(JSONObject(c.toString()).put("red", tipo))
                continue
            }
            val nombre = c.getString("etiqueta")
            var primera = true
            for (tipo in listOf("kda", "evm")) {
                val cu = porTipo[tipo] ?: continue
                val id = if (primera) c.getString("id") else libre(usados)
                primera = false
                usados.add(id)
                nuevas.put(JSONObject()
                    .put("id", id)
                    .put("etiqueta", nombre + " " + tipo.uppercase())
                    .put("semilla", c.getString("semilla"))
                    .put("red", tipo)
                    .put("cuentas", JSONArray().put(JSONObject(cu.toString()).put("id", "$id-$tipo"))))
            }
        }
        val activa = datos.optString("activa")
        return JSONObject()
            .put("v", VERSION)
            .put("carteras", nuevas)
            .put("activa", if (nuevas.length() == 0) "" else activa.ifEmpty { nuevas.getJSONObject(0).getString("id") })
    }

    private fun libre(usados: Set<String>): String {
        var n = 1
        while (usados.contains("c$n")) n++
        return "c$n"
    }

    fun lista(datos: JSONObject): JSONArray = normalizar(datos).getJSONArray("carteras")

    fun buscar(datos: JSONObject, id: String): JSONObject? {
        val l = lista(datos)
        for (i in 0 until l.length()) {
            val c = l.getJSONObject(i)
            if (c.getString("id") == id) return c
        }
        return null
    }

    /** La cartera activa, o la primera si la marcada ya no existe. */
    fun activa(datos: JSONObject): JSONObject {
        val d = normalizar(datos)
        return buscar(d, d.optString("activa")) ?: lista(d).getJSONObject(0)
    }

    /** Identificador libre: c1, c2, c3... Nunca se reutiliza uno ya usado. */
    fun idNuevo(datos: JSONObject): String {
        val l = lista(datos)
        var n = 1
        val usados = mutableSetOf<String>()
        for (i in 0 until l.length()) usados.add(l.getJSONObject(i).getString("id"))
        while (usados.contains("c$n")) n++
        return "c$n"
    }

    fun anadir(datos: JSONObject, cartera: JSONObject): JSONObject {
        val d = normalizar(datos)
        d.getJSONArray("carteras").put(cartera)
        d.put("activa", cartera.getString("id"))     // la recien creada pasa a ser la activa
        return d
    }

    /**
     * Quita una cartera. Se niega a quitar la ultima: una boveda sin carteras
     * dejaria la app en un estado raro -ni primer uso ni cartera- y lo que quiere
     * quien borra la unica que tiene es empezar de cero, que es otra cosa
     * (`borrarTodo`) y avisa de lo que hace.
     */
    fun quitar(datos: JSONObject, id: String): JSONObject {
        val d = normalizar(datos)
        val l = d.getJSONArray("carteras")
        if (l.length() <= 1) throw IllegalStateException("Es la única cartera que hay. Para empezar de cero, usa borrar todo.")
        val nueva = JSONArray()
        for (i in 0 until l.length()) {
            val c = l.getJSONObject(i)
            if (c.getString("id") != id) nueva.put(c)
        }
        if (nueva.length() == l.length()) throw IllegalArgumentException("Esa cartera no existe.")
        d.put("carteras", nueva)
        if (d.optString("activa") == id) d.put("activa", nueva.getJSONObject(0).getString("id"))
        return d
    }

    fun renombrar(datos: JSONObject, id: String, etiqueta: String): JSONObject {
        val d = normalizar(datos)
        val c = buscar(d, id) ?: throw IllegalArgumentException("Esa cartera no existe.")
        val limpia = etiqueta.trim().take(40)
        if (limpia.isEmpty()) throw IllegalArgumentException("Ponle un nombre.")
        c.put("etiqueta", limpia)
        return d
    }

    /**
     * ¿Ya hay una cartera con esta misma semilla PARA ESA RED? Evita duplicados
     * sin sentido, pero permite lo que ahora es legitimo: la misma semilla en dos
     * carteras, una de Kadena y otra de Ethereum. De hecho es exactamente lo que
     * deja la migracion v2 -> v3.
     */
    fun yaExiste(datos: JSONObject, semilla: String, red: String): Boolean {
        val l = lista(datos)
        for (i in 0 until l.length()) {
            val c = l.getJSONObject(i)
            // optString: las carteras metidas por su clave privada no tienen semilla.
            if (c.optString("semilla") == semilla && redDe(c) == red) return true
        }
        return false
    }

    /** ¿Ya esta metida esta clave privada para esa red? */
    fun yaExisteClave(datos: JSONObject, privada: String, red: String): Boolean {
        val l = lista(datos)
        for (i in 0 until l.length()) {
            val c = l.getJSONObject(i)
            if (c.optString("privada") == privada && redDe(c) == red) return true
        }
        return false
    }

    /**
     * ¿Hay ya una cartera con ESA MISMA direccion? Importar la clave privada de
     * una cartera que ya esta por su semilla dejaria dos entradas con la misma
     * cuenta y el mismo dinero: confuso y sin ninguna ventaja.
     */
    fun yaExisteCuenta(datos: JSONObject, cuenta: String): Boolean {
        val l = lista(datos)
        for (i in 0 until l.length()) {
            val cuentas = l.getJSONObject(i).optJSONArray("cuentas") ?: continue
            for (j in 0 until cuentas.length()) {
                if (cuentas.getJSONObject(j).optString("cuenta") == cuenta) return true
            }
        }
        return false
    }

    /** ¿Esta cartera viene de unas palabras, o de una clave suelta? */
    fun esDeSemilla(cartera: JSONObject): Boolean = cartera.optString("semilla").isNotEmpty()

    /**
     * Deja la clave privada como se guarda: 64 caracteres hex en minusculas.
     *
     * Se admite el `0x` de delante -las de Ethereum se copian asi de todas partes-
     * y tambien el formato de 128 caracteres de algunas herramientas de Kadena,
     * que pegan la privada y la publica seguidas: la publica se recalcula, asi que
     * la segunda mitad sobra... pero solo si de verdad es la publica de esa
     * privada. Si no cuadra, no se acepta: sera otra cosa.
     */
    fun normalizarClave(texto: String, red: String): String {
        var h = texto.trim().removePrefix("0x").removePrefix("0X").lowercase()
        if (!Regex("^[0-9a-f]+$").matches(h)) {
            throw IllegalArgumentException("Una clave privada son 64 caracteres del 0 al 9 y de la a a la f.")
        }
        if (h.length == 128 && red == "kda") {
            val privada = h.substring(0, 64)
            val publica = h.substring(64)
            val calculada = Derivacion.publicaKadena(Derivacion.deHex(privada))
            if (calculada != publica) {
                throw IllegalArgumentException("Esos 128 caracteres no son una clave privada seguida de su pública.")
            }
            h = privada
        }
        if (h.length != 64) {
            throw IllegalArgumentException("Una clave privada son 64 caracteres del 0 al 9 y de la a a la f.")
        }
        if (h == "0".repeat(64)) throw IllegalArgumentException("Esa clave no vale: son todo ceros.")
        return h
    }

    /**
     * Monta una cartera a partir de una clave privada suelta, sin semilla.
     *
     * Esa cartera NO tiene palabras y no se pueden inventar: quien la meta asi
     * solo podra sacar de aqui su clave privada, y la pantalla lo dice.
     */
    fun montarConClave(id: String, etiqueta: String, privada: String, red: String): JSONObject {
        if (red != "kda" && red != "evm") throw IllegalArgumentException("Esa red no existe.")
        val bytes = Derivacion.deHex(privada)
        val cuenta = try {
            if (red == "kda") {
                JSONObject().put("id", "$id-kda").put("etiqueta", "Kadena").put("tipo", "kda")
                    .put("cuenta", "k:" + Derivacion.publicaKadena(bytes))
            } else {
                JSONObject().put("id", "$id-evm").put("etiqueta", "EVM").put("tipo", "evm")
                    .put("cuenta", Derivacion.direccionEvm(bytes))
            }
        } finally {
            bytes.fill(0)
        }
        return JSONObject()
            .put("id", id).put("etiqueta", etiqueta).put("privada", privada).put("red", red)
            .put("cuentas", JSONArray().put(cuenta))
    }

    /**
     * Los 32 bytes de la clave privada de esa cartera, venga de donde venga.
     *
     * Quien la pida se la lleva en un array que puede -y debe- borrar despues.
     */
    fun privadaDe(cartera: JSONObject): ByteArray {
        val suelta = cartera.optString("privada")
        if (suelta.isNotEmpty()) return Derivacion.deHex(suelta)
        val bytes = Derivacion.semillaABytes(cartera.getString("semilla"))
        return try {
            if (redDe(cartera) == "kda") Derivacion.privadaKadena(bytes, 0)
            else Derivacion.privadaEvm(bytes, 0)
        } finally {
            bytes.fill(0)
        }
    }

    /** La red de una cartera. Las de antes de la v3 no la traen: se deduce. */
    fun redDe(cartera: JSONObject): String {
        val r = cartera.optString("red")
        if (r == "kda" || r == "evm") return r
        val cuentas = cartera.optJSONArray("cuentas") ?: return "kda"
        return if (cuentas.length() > 0) cuentas.getJSONObject(0).optString("tipo", "kda") else "kda"
    }

    /**
     * Monta una cartera nueva a partir de una semilla, derivando SU cuenta.
     *
     * Una cartera, una red. La semilla sirve para las dos -son dos caminos BIP-44
     * distintos- asi que quien quiera las dos crea dos carteras con las mismas
     * palabras, y la app no se lo impide.
     */
    fun montar(id: String, etiqueta: String, semilla: String, red: String): JSONObject {
        if (red != "kda" && red != "evm") throw IllegalArgumentException("Esa red no existe.")
        val bytes = Derivacion.semillaABytes(semilla)
        val cuenta = try {
            if (red == "kda") {
                JSONObject().put("id", "$id-kda").put("etiqueta", "Kadena").put("tipo", "kda")
                    .put("cuenta", Derivacion.cuentaKadena(bytes, 0))
            } else {
                JSONObject().put("id", "$id-evm").put("etiqueta", "EVM").put("tipo", "evm")
                    .put("cuenta", Derivacion.direccionEvm(Derivacion.privadaEvm(bytes, 0)))
            }
        } finally {
            bytes.fill(0)
        }
        return JSONObject()
            .put("id", id).put("etiqueta", etiqueta).put("semilla", semilla).put("red", red)
            .put("cuentas", JSONArray().put(cuenta))
    }
}
